File: src/2023/11.py
#!/usr/bin/env python

import copy

FILENAME = "../../data/2023/11.txt"


class Galaxy:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

    def distance(self, galaxy: "Galaxy") -> int:
        return abs(self.x - galaxy.x) + abs(self.y - galaxy.y)


class Universe:
    def __init__(self, data: list[str]):
        self.galaxies = []
        self.height = len(data)
        self.width = len(data[0])

        for y, line in enumerate(data):
            for x, char in enumerate(line):
                if char == "#":
                    self.galaxies.append(Galaxy(x, y))

    def expand(self, amount: int = 2) -> None:
        x_voids = set(range(self.width))
        y_voids = set(range(self.height))

        for g in self.galaxies:
            x_voids.discard(g.x)
            y_voids.discard(g.y)

        new_galaxies = copy.deepcopy(self.galaxies)
        new_width = self.width
        new_height = self.height
        expansion = amount - 1

        for i, galaxy in enumerate(self.galaxies):
            for x in x_voids:
                if galaxy.x > x:
                    new_galaxies[i].x += expansion
                    new_width = max(new_width, new_galaxies[i].x)

            for y in y_voids:
                if galaxy.y > y:
                    new_galaxies[i].y += expansion
                    new_height = max(new_height, new_galaxies[i].y)

        self.galaxies = copy.deepcopy(new_galaxies)
        self.width = new_width
        self.height = new_height

    def __repr__(self) -> str:
        u = []
        for y in range(self.height):
            u.append([])
            for x in range(self.width):
                u[y].append(".")
            u.append("\n")
        u.append("\n")

        for g in self.galaxies:
            u[g.y][g.x] = "#"

        return "".join(
            [u[y][x] for x in range(self.height) for y in range(self.width)]
        )


def part_1(data: list[str]) -> int:
    return part_2(data, 2)


def part_2(data: list[str], amount: int) -> int:
    universe = Universe(data)
    universe.expand(amount)

    total_distance = 0
    n = len(universe.galaxies)
    for i in range(n):
        for j in range(i + 1, n):
            distance = universe.galaxies[i].distance(universe.galaxies[j])
            total_distance += distance

    return total_distance



if __name__ == "__main__":
    test_data = [
        "...#......",
        ".......#..",
        "#.........",
        "..........",
        "......#...",
        ".#........",
        ".........#",
        "..........",
        ".......#..",
        "#...#.....",
    ]

    total_distance = part_1(test_data)
    assert total_distance == 374

    total_distance = part_2(test_data, 100)
    assert total_distance == 8410

    with open(FILENAME, "r") as f:
        data = [line.rstrip("\n") for line in f]

    print(f"Part 1: {part_1(data)}")
    print(f"Part 2: {part_2(data, 1000000)}")

File: src/2023/src/11.cpp
#include <set>

#include "11.hpp"

using namespace std;
using namespace myg;

Universe::Universe(const vector<string> data) {
    _height = data.size();
    _width = data[0].size();
    _galaxies = vector<galaxy>();

    for (auto i = ZERO; i < _height; i++)
        for (auto j = ZERO; j < _width; j++) {
            if (data[i][j] == '#') {
                const uint x = static_cast<uint>(i);
                const uint y = static_cast<uint>(j);
                struct galaxy g = { x, y };

                _galaxies.push_back(g);
            }
        }

    _n = _galaxies.size();
}

void Universe::expand(const uint amount) {
    set<uint> x_voids = set<uint>();
    for (auto i = ZERO; i < _width; i++)
        x_voids.insert(static_cast<uint>(i));

    set<uint> y_voids = set<uint>();
    for (auto i = ZERO; i < _height; i++)
        y_voids.insert(static_cast<uint>(i));

    for (auto g: _galaxies) {
        x_voids.erase(g.x);
        y_voids.erase(g.y);
    }

    vector<galaxy> new_galaxies = vector<galaxy>(_galaxies);
    uint new_width = static_cast<uint>(_width);
    uint new_height = static_cast<uint>(_height);
    const uint expansion = amount - 1;

    for (auto i = ZERO; i < _n; i++) {
        for (auto x: x_voids)
            if (_galaxies[i].x > x) {
                new_galaxies[i].x += expansion;
                new_width = max(new_width, new_galaxies[i].x);
            }

        for (auto y: y_voids)
            if (_galaxies[i].y > y) {
                new_galaxies[i].y += expansion;
                new_height = max(new_height, new_galaxies[i].y);
            }
    }

    _galaxies = new_galaxies;
    _width = static_cast<size_t>(new_width);
    _height = static_cast<size_t>(new_height);
};

ulong Universe::total_distance(void) const {
    ulong d_total = 0;

    for (auto i = ZERO; i < _n; i++)
        for (auto j = i + 1; j < _n; j++)
            d_total += distance(i, j);

    return d_total;
};

ulong Universe::distance(const size_t i, const size_t j) const {
    const uint d_horizontal =
        _galaxies[i].x < _galaxies[j].x ?
        _galaxies[j].x - _galaxies[i].x :
        _galaxies[i].x - _galaxies[j].x;

    const uint d_vertical =
        _galaxies[i].y < _galaxies[j].y ?
        _galaxies[j].y - _galaxies[i].y :
        _galaxies[i].y - _galaxies[j].y;

    return static_cast<ulong>(d_horizontal) + static_cast<ulong>(d_vertical);
};

ulong Day11::part_1(void) {
    return part_2(2);
};

ulong Day11::part_2(const uint amount) {
    Universe u = Universe(_data);
    u.expand(amount);
    return u.total_distance();
};
